runners/flink/src/main/java/org/apache/beam/runners/flink/LyftFlinkStreamingPortableTranslations.java

```diff
@@ -192,13 +235,117 @@ public KinesisByteArrayWindowedValueSchema() {
     }
 
     @Override
-    public WindowedValue<byte[]> deserialize(byte[] recordValue) {
+    public WindowedValue<byte[]> deserialize(
+        byte[] recordValue,
+        String partitionKey,
+        String seqNum,
+        long approxArrivalTimestamp,
+        String stream,
+        String shardId)
+        throws IOException {
       return WindowedValue.valueInGlobalWindow(recordValue);
     }
+  }
+
+  /**
+   * Deserializer for Lyft's kinesis event format, which is a zlib-compressed, base64-encoded, json
+   * array of event objects. This schema tags events with the occurred_at time of the oldest event
+   * in the message.
+   *
+   * <p>The output of this schema is utf-8 encoded json.
+   */
+  private static class LyftBeamKinesisSchema
+      implements KinesisDeserializationSchema<WindowedValue<byte[]>> {
+    private static final ObjectMapper mapper = new ObjectMapper();
+    private static final TypeInformation<WindowedValue<byte[]>> ti =
+        new CoderTypeInformation<>(
+            WindowedValue.getFullCoder(ByteArrayCoder.of(), GlobalWindow.Coder.INSTANCE));
+
+    private static String inflate(byte[] deflatedData) throws IOException {
+      Inflater inflater = new Inflater();
+      inflater.setInput(deflatedData);
+
+      // buffer for inflating data
+      byte[] buf = new byte[4096];
+
+      try (ByteArrayOutputStream bos = new ByteArrayOutputStream(deflatedData.length)) {
+        while (!inflater.finished()) {
+          int count = inflater.inflate(buf);
+          bos.write(buf, 0, count);
+        }
+        return bos.toString("UTF-8");
+      } catch (DataFormatException e) {
+        throw new IOException("Failed to expand message", e);
+      }
+    }
+
+    private static long parseDateTime(String datetime) {
+      try {
+        DateTimeFormatter formatterToUse =
+            (datetime.length() - datetime.indexOf('.') == 7)
+                ? DB_DATETIME_FORMATTER
+                : BACKUP_DB_DATETIME_FORMATTER;
+        TemporalAccessor temporalAccessor = formatterToUse.parse(datetime);
+        LocalDateTime localDateTime = LocalDateTime.from(temporalAccessor);
+        ZonedDateTime zonedDateTime = ZonedDateTime.of(localDateTime, GMT);
+        return java.time.Instant.from(zonedDateTime).toEpochMilli();
+      } catch (DateTimeParseException e) {
+        return java.time.Instant.from(ISO_DATETIME_FORMATTER.parse(datetime)).toEpochMilli();
+      }
+    }
 
     @Override
-    public boolean isEndOfStream(WindowedValue<byte[]> nextElement) {
-      return false;
+    public WindowedValue<byte[]> deserialize(
+        byte[] recordValue,
+        String partitionKey,
+        String seqNum,
+        long approxArrivalTimestamp,
+        String stream,
+        String shardId)
+        throws IOException {
+      String inflatedString = inflate(recordValue);
+
+      JsonNode events = mapper.readTree(inflatedString);
```

Comment:
Just a thought, instead of parsing the JSON, could we do this regex based? This would also allow us to make this more configurable.

Reply:
The idea of "parsing" json using a regex scares me a bit. I worry that might be overly sensitive to changes in how we produce the json. I think it's more reliable to use a real json parser, even if there's a bit of extra overhead from that.

Reply:
Thinking about this a bit more, there are basically three things the client needs to describe about their data format:

1. How to decode the data into a java object (i.e., JsonNode in this case)
2. How to access the timestamp data there ("occurred_at")
3. How to parse the timestamp

Across all Lyft use cases I'd expect only a few different versions of 1 (i.e., in kafka it would probably be just zlib-compressed, but not base64-encoded; we'll also have protobuf formats), so it's feasible to just implement those here. 2 will probably be diverse but hopefully easy to specify from python as a string (the main complexity here is handling arrays of events vs. invidual events). 3 should also be able to be handled by a small number of date formats.

For now, I think it simplest to just hard-code the logic for fanner events, but it's definitely good to start thinking about how we can generalize this.

Reply:
Which timestamp to use is what will vary most across use cases. That's why I was thinking about a more flexible option to extract it. Normally you would want to specify a lambda for this, but that's not an option here (Python). Something that takes an expression to extract from JSON, ideally even without having to construct a full JSON object tree, would be nice.